### PR TITLE
trans rewrite

### PR DESCRIPTION
**Pull request by sanderclaeys**
I rewrote the transformer methods as preparation for the convex relaxations of transformers and OLTC. Like this the ACP formulation is more readable, and it will fit the relaxations better as well.

**The dispatch to OLTC or fixed transformer might need some discussion.**
For me, the default problem should include OLTCs. If the user does not want OLTCs, she should fix the taps in the data model. The formulations should then resort automatically to the simplified equations instead for fixed taps. Last time, I remember there was some preference for ignoring the `fixed` setting and instead considering it as fixed in most problem definitions. I considered both by implementing the following logic.
- `constraint_mc_oltc` in the constraint template calls `constraint_mc_trans_yy` or `constraint_mc_trans_dy` depending on the winding types, and passes the `fixed` parameter of the data model. These methods are OLTC ready, and can simplify the equations when all/some of the taps are fixed.
- `constraint_mc_trans` is identical to `constraint_mc_oltc`, but ignores the data model and passes a vector of ones, indicating all taps are fixed to `constraint_mc_trans_yy` and `constraint_mc_trans_dy`.

This achieves a good balance between code duplication, readability and exploiting fixed variables to reduce the complexity where possible.

### CODE REVIEW

File: src/core/constraint_template.jl [hunk 1]
```diff
@@ -134,40 +134,83 @@ function constraint_mc_flow_losses(pm::_PMs.AbstractPowerModel, i::Int; nw::Int=
...
+        constraint_mc_trans_yy(pm, nw, i, f_bus, t_bus, f_idx, t_idx, f_cnd, t_cnd, pol, tm_set, tm_fixed, tm_scale)
+    elseif f_type=="delta" && t_type=="wye"
+        constraint_mc_trans_dy(pm, nw, i, f_bus, t_bus, f_idx, t_idx, f_cnd, t_cnd, pol, tm_set, tm_fixed, tm_scale)
+    elseif f_type=="wye" && t_type=="delta"
+        constraint_mc_trans_dy(pm, nw, i, t_bus, f_bus, t_idx, f_idx, t_cnd, f_cnd, pol, tm_set, tm_fixed, (tm_scale)^-1)
+    elseif f_type=="delta" && t_type=="delta"
+        Memento.error(_LOGGER, "Dd transformers are not supported at the low-level data format. This can be cast as a combo of two dy transformers.")
+    end
+    if f_type=="zig-zag" || t_type=="zig-zag"
+        Memento.error("Zig-zag not yet supported.")
```
Comment (pseudocubic) [inline src/core/constraint_template.jl]: Missing `_LOGGER` => `Memento.error(_LOGGER, "Zig-zag not yet supported.")`

File: src/core/constraint_template.jl [hunk 2]
```diff
@@ -134,40 +134,83 @@ function constraint_mc_flow_losses(pm::_PMs.AbstractPowerModel, i::Int; nw::Int=
...
```
Comment (pseudocubic) [inline src/core/constraint_template.jl]: Missing `_LOGGER` => `Memento.error(_LOGGER, "Zig-zag not yet supported.")`

File: src/core/data.jl [hunk 3]
```diff
@@ -113,3 +113,30 @@ function count_nodes(pmd_data::Dict{String,Any})::Int
...
+        tm_scale *= sqrt(3)
+    end
+    if t_type == "delta"
+        tm_scale *= 1/sqrt(3)
+    end
+    if f_type == "zig-zag"
+        Memento.error("Zig-zag not yet supported.")
+    end
+    if t_type == "zig-zag"
+        Memento.error("Zig-zag not yet supported.")
```
Comment (pseudocubic) [inline src/core/data.jl]: Missing `_LOGGER` => `Memento.error(_LOGGER, "Zig-zag not yet supported.")`

File: src/core/data.jl [hunk 4]
```diff
@@ -113,3 +113,30 @@ function count_nodes(pmd_data::Dict{String,Any})::Int
...
+
+    tm_scale = (f_vnom/t_vnom)*(t_vbase/f_vbase)
+    if f_type == "delta"
+        tm_scale *= sqrt(3)
+    end
+    if t_type == "delta"
+        tm_scale *= 1/sqrt(3)
+    end
+    if f_type == "zig-zag"
+        Memento.error("Zig-zag not yet supported.")
```
Comment (pseudocubic) [inline src/core/data.jl]: Missing `_LOGGER` => `Memento.error(_LOGGER, "Zig-zag not yet supported.")`

File: src/core/constraint_template.jl [hunk 5]
```diff
@@ -134,40 +134,83 @@ function constraint_mc_flow_losses(pm::_PMs.AbstractPowerModel, i::Int; nw::Int=
...
+    end
 end
```
Comment (pseudocubic) [inline src/core/constraint_template.jl]: The differences between these two functions seems pretty minimal, is it possible to combine them with a OTLC keyword argument, e.g. `function constraint_mc_trans(pm::_PMs.AbstractPowerModel, i::Int; nw::Int=pm.cnw, oltc::Bool=false)`?

File: src/core/variable.jl [hunk 6]
```diff
@@ -190,18 +190,18 @@ end
...
 function variable_mc_oltc_tap(pm::_PMs.AbstractPowerModel; nw::Int=pm.cnw, bounded=true)
-    nphases = 3
-    oltc_ids = _PMs.ids(pm, pm.cnw, :transformer)
-    for c in 1:nphases
-        _PMs.var(pm, nw, c)[:tap] = JuMP.@variable(pm.model,
-            [i in oltc_ids],
-            base_name="$(nw)_tm",
-            start=_PMs.ref(pm, nw, :transformer, i, "tm")[c]
+    nph = 3
+    for p in 1:nph
```
Comment (pseudocubic) [inline src/core/variable.jl]: should loop like `for p in _PMs.conductor_ids(pm)` for easier extensibility. If you want to fix to exactly three phases I would use something like `@assert length(_PMs.conductor_ids(pm) == 3`

File: src/form/acp.jl [hunk 7]
```diff
@@ -334,73 +334,91 @@ function constraint_mc_transformer_flow(pm::_PMs.AbstractACPModel, nw::Int, i::I
...
+
+    vm_fr = [_PMs.var(pm, nw, p, :vm, f_bus) for p in f_cnd]
+    vm_to = [_PMs.var(pm, nw, p, :vm, t_bus) for p in t_cnd]
+    va_fr = [_PMs.var(pm, nw, p, :va, f_bus) for p in f_cnd]
+    va_to = [_PMs.var(pm, nw, p, :va, t_bus) for p in t_cnd]
+
+    # construct tm as a parameter or scaled variable depending on whether it is fixed or not
+    tm = [tm_fixed[p] ? tm_set[p] : _PMs.var(pm, nw, p, :tap, trans_id) for p in 1:nph]
+
+    for p in 1:nph
```
Comment (pseudocubic) [inline src/form/acp.jl]: I think using `_PMs.conductor_ids(pm)` would be more appropriate here

File: src/form/acp.jl [hunk 8]
```diff
@@ -334,73 +334,91 @@ function constraint_mc_transformer_flow(pm::_PMs.AbstractACPModel, nw::Int, i::I
...
+    q_fr = [_PMs.var(pm, nw, p, :qt, f_idx) for p in f_cnd]
+    q_to = [_PMs.var(pm, nw, p, :qt, t_idx) for p in t_cnd]
+
+    id_re = Array{Any,1}(undef, nph)
+    id_im = Array{Any,1}(undef, nph)
+    # s/v      = (p+jq)/|v|^2*conj(v)
+    #          = (p+jq)/|v|*(cos(va)-j*sin(va))
+    # Re(s/v)  = (p*cos(va)+q*sin(va))/|v|
+    # -Im(s/v) = -(q*cos(va)-p*sin(va))/|v|
+    for p in 1:nph
```
Comment (pseudocubic) [inline src/form/acp.jl]: I think using `_PMs.conductor_ids(pm)` would be more appropriate here

File: src/form/acp.jl [hunk 9]
```diff
@@ -334,73 +334,91 @@ function constraint_mc_transformer_flow(pm::_PMs.AbstractACPModel, nw::Int, i::I
...
+        # id = conj(s_to/v_to)./tm
+        id_re[p] = JuMP.@NLexpression(pm.model,  (p_to[p]*cos(va_to[p])+q_to[p]*sin(va_to[p]))/vm_to[p]/(tm_scale*tm[p])/pol)
+        id_im[p] = JuMP.@NLexpression(pm.model, -(q_to[p]*cos(va_to[p])-p_to[p]*sin(va_to[p]))/vm_to[p]/(tm_scale*tm[p])/pol)
+    end
+    for p in 1:nph
```
Comment (pseudocubic) [inline src/form/acp.jl]: I think using `_PMs.conductor_ids(pm)` would be more appropriate here

File: src/core/constraint_template.jl (hunk 5 above)
Comment (sanderclaeys) [reply src/core/constraint_template.jl]: Yes most definitely, I kept this mostly not to break too much with the old naming scheme.

File: src/core/variable.jl (hunk 6 above)
Comment (sanderclaeys) [reply src/core/variable.jl]: This was mostly done with the 4-wire scheme in mind, where phases would be different from conductors. But that would require further editing in any case, so I will implement it as you suggest everywhere for now.

**Comment by pseudocubic** [comment]
@sanderclaeys to address the point you bring up for discussion about whether oltc should be specified when instantiating a problem definition or by the data model, I personally am inclined towards your preference of specifying this in the data. As long as this feature is well documented I don't think there should be any problem. 

That being said, I think that this issue should be left to a separate PR after this and #150 are resolved, when we revisit the problem definitions in the very near future

**Comment by sanderclaeys** [comment]
Sounds good, I will create an issue for this as a reminder. 

Also, the transformer code now implicitly assumes that each conductor is also a phase; this code will break for 4-wire systems, where the neutral is part of `_PMs.conductor_ids(pm)`. This should be easy enough to update once we decide on a pattern for neutral vs. phase conductors.

**Pull request merged**